Firestore/Example/Tests/GoogleTest/FSTGoogleTestTests.mm

```diff
@@ -0,0 +1,181 @@
+  NSString *testClassName = NSStringFromClass([self class]);
+  NSUInteger end = testClassName.length - 5;
+  const char *testCaseName = [[testClassName substringToIndex:end] UTF8String];
+
+  // Find the googletest TestInfo name by removing the leading "test"
+  NSString *selectorName = NSStringFromSelector(_cmd);
+  const char *testInfoName = [[selectorName substringFromIndex:4] UTF8String];
+
+  // TODO(wilhuff): Avoid linear search here because it makes test execution overall O(n^2)
+  const testing::UnitTest *master = testing::UnitTest::GetInstance();
+  int testCases = master->total_test_case_count();
+  for (int i = 0; i < testCases; i++) {
+    const testing::TestCase *testCase = master->GetTestCase(i);
+    if (strcmp(testCaseName, testCase->name()) != 0) {
+      continue;
+    }
+
+    int testInfos = testCase->total_test_count();
+    for (int j = 0; j < testInfos; j++) {
+      const testing::TestInfo *testInfo = testCase->GetTestInfo(j);
+      if (strcmp(testInfoName, testInfo->name()) == 0) {
+        return testInfo;
+      }
+    }
+  }
+
+  XCTFail(@"Failed to find a TestInfo corresponding to class %@ method %@", testClassName,
+          selectorName);
+  return nullptr;
+}
+
+/** Finds the TestInfo corresponding to this test and reports it to XCTest */
+void ReportTestResult(XCTestCase *self, SEL _cmd) {
+  const testing::TestInfo *testInfo = FindTestInfo(self, _cmd);
+  if (!testInfo) {
+    return;
+  }
+
+  if (!testInfo->should_run()) {
+    // Test was filtered out by gunit; nothing to report.
+    return;
+  }
+
+  const testing::TestResult *result = testInfo->result();
+  if (result->Passed()) {
+    // Let XCode know that the test ran and succeeded.
+    XCTAssertTrue(true);
+    return;
+  }
+
+  // Test failed :-(. Record the failure such that XCode will navigate directly to the file:line.
+  int parts = result->total_part_count();
+  for (int i = 0; i < parts; i++) {
+    const testing::TestPartResult &part = result->GetTestPartResult(i);
+
+    NSString *message = [NSString stringWithCString:part.message() encoding:NSUTF8StringEncoding];
+    NSString *fileName = @"";
+    if (part.file_name()) {
+      fileName = [NSString stringWithCString:part.file_name() encoding:NSUTF8StringEncoding];
+    }
+    NSUInteger lineNumber = 0;
+    if (part.line_number() > 0) {
+      lineNumber = part.line_number();
+    }
+
+    [self recordFailureWithDescription:message inFile:fileName atLine:lineNumber expected:YES];
+  }
+}
+
+/**
+ * Generates a new subclass of XCTestCase for the given GoogleTest TestCase. Each TestInfo (which
+ * represents an indivudal test method execution) is translated into a method on the test case.
+ *
+ * @return A new Class that's a subclass of XCTestCase, that's been registered with the Objective-C
+ *     runtime.
+ */
+Class CreateXCTestCaseClass(const testing::TestCase *testCase) {
+  NSString *testCaseName = [NSString stringWithFormat:@"%sTests", testCase->name()];
```

Comment:
Moving "Tests" to a constant string may be handy here and in the code that strips the suffix above.

Reply:
Done.